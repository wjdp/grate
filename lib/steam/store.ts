import { z } from "zod";

export class SteamStoreError extends Error {
  retriable: boolean;
  constructor(message: string, retriable = false) {
    super(message);
    this.name = "SteamStoreError";
    this.retriable = retriable;
  }
}

const SteamAppInfo = z.object({
  type: z.string(),
  name: z.string(),
  steam_appid: z.number(),
  required_age: z.union([z.number(), z.string()]),
  is_free: z.boolean(),
  detailed_description: z.string(),
  about_the_game: z.string(),
  short_description: z.string(),
  header_image: z.string(),
  capsule_image: z.string(),
  capsule_imagev5: z.string(),
  website: z.string().nullable(),
  developers: z.array(z.string()),
  publishers: z.array(z.string()).optional(),
  platforms: z.object({
    windows: z.boolean(),
    mac: z.boolean(),
    linux: z.boolean(),
  }),
  metacritic: z
    .object({
      score: z.number(),
      url: z.string().optional(),
    })
    .optional(),
  categories: z
    .array(
      z.object({
        id: z.number(),
        description: z.string(),
      }),
    )
    .optional(),
  genres: z
    .array(
      z.object({
        id: z.string(),
        description: z.string(),
      }),
    )
    .optional(),
  screenshots: z
    .array(
      z.object({
        id: z.number(),
        path_thumbnail: z.string(),
        path_full: z.string(),
      }),
    )
    .optional(),
  release_date: z
    .object({
      coming_soon: z.boolean(),
      date: z.string(),
    })
    .optional(),
  background: z.string(),
  background_raw: z.string(),
});

export async function getAppDetails(appId: number) {
  const response = await fetch(
    `http://store.steampowered.com/api/appdetails/?appids=${appId}`,
  );
  if (!response.ok) {
    if (response.status === 429) {
      throw new SteamStoreError("Too many requests", true);
    }
    throw new SteamStoreError(response.statusText, true);
  }
  const data = await response.json();
  const appData = data[appId.toString()];
  if (!appData.success) {
    throw new SteamStoreError("App details unavailable", false);
  }
  return SteamAppInfo.parse(appData.data);
}

export function parseReleaseDate(s: string) {
  if (!s) {
    return null;
  }
  // date from steam is like: 25 Mar, 2013
  let date;
  try {
    date = new Date(s);
  } catch (error) {
    console.error(`Error parsing date: ${date}`);
    throw error;
  }
  if (isNaN(date.getTime())) {
    throw new Error(`Invalid date: ${date}`);
  }
  return date;
}
