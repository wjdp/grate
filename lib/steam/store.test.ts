import createFetchMock from "vitest-fetch-mock";
import { vi, beforeEach, describe, expect, it, afterAll } from "vitest";
import { getAppDetails, parseReleaseDate, SteamStoreError } from "./store";

import Response7670 from "./fixtures/store/7670.json";
import Response443080 from "./fixtures/store/443080.json";

const BIOSHOCK_APP_ID = 7670;
const INVALID_APP_ID = 443080;

const fetchMocker = createFetchMock(vi);
fetchMocker.enableMocks();

describe("getAppDetails", () => {
  beforeEach(() => {
    fetchMocker.resetMocks();
  });
  afterAll(() => {
    fetchMocker.disableMocks();
  });
  it("should return a SteamAppInfo object for a valid app ID", async () => {
    fetchMocker.mockResponseOnce(JSON.stringify(Response7670));
    const appDetails = await getAppDetails(BIOSHOCK_APP_ID);
    expect(appDetails).toHaveProperty("name", "BioShockâ„¢");
    expect(appDetails).toMatchSnapshot();
  });
  it("should handle an invalid app ID", async () => {
    fetchMocker.mockResponse(JSON.stringify(Response443080));
    await expect(getAppDetails(INVALID_APP_ID)).rejects.toThrow(
      SteamStoreError,
    );
    try {
      await getAppDetails(INVALID_APP_ID);
    } catch (error) {
      if (error instanceof SteamStoreError) {
        expect(error.retriable).toBe(false);
      } else {
        throw error;
      }
    }
  });
  it("should handle too many requests", async () => {
    fetchMocker.mockResponse("", { status: 429 });
    await expect(getAppDetails(BIOSHOCK_APP_ID)).rejects.toThrowError(
      SteamStoreError,
    );
    try {
      await getAppDetails(INVALID_APP_ID);
    } catch (error) {
      if (error instanceof SteamStoreError) {
        expect(error.retriable).toBe(true);
      } else {
        throw error;
      }
    }
  });
});

describe("parseReleaseDate", () => {
  it("should return a Date object for a valid steam release date", () => {
    const date = parseReleaseDate("25 Mar, 2013");
    if (!date) {
      throw new Error("Expected a date object");
    }
    expect(date).toBeInstanceOf(Date);
    expect(date.getFullYear()).toBe(2013);
    expect(date.getMonth()).toBe(2);
    expect(date.getDate()).toBe(25);
  });
  it("throws on invalid date", async () => {
    expect(() => parseReleaseDate("invalid date")).toThrow();
  });
  it("handles empty string", () => {
    const date = parseReleaseDate("");
    expect(date).toBe(null);
  });
});
