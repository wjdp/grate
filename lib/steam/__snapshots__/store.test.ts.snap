// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAppDetails > should return a SteamAppInfo object for a valid app ID 1`] = `
{
  "about_the_game": "BioShock is a shooter unlike any you've ever played, loaded with weapons and tactics never seen. You'll have a complete arsenal at your disposal from simple revolvers to grenade launchers and chemical throwers, but you'll also be forced to genetically modify your DNA to create an even more deadly weapon: you. Injectable plasmids give you super human powers: blast electrical currents into water to electrocute multiple enemies, or freeze them solid and obliterate them with the swing of a wrench.  <br>No encounter ever plays out the same, and no two gamers will play the game the same way.<br><br><ul class="bb_ul"><li>Biologically modify your body: send fire storming from your fingertips and unleash a swarm of killer hornets hatched from the veins in your arms.<br></li><li>Hack devices and systems, upgrade your weapons and craft new ammo variants.<br></li><li>Turn everything into a weapon: the environment, your body, fire and water, and even your worst enemies.<br></li><li>Explore an incredible and unique art deco world hidden deep under the ocean.</li></ul><h2 class="bb_tag">BioShock: Breaking the Mold</h2>A free download that takes an inside look at the art of BioShock.",
  "background": "https://store.akamai.steamstatic.com/images/storepagebackground/app/7670?t=1730172622",
  "background_raw": "https://store.akamai.steamstatic.com/images/storepagebackground/app/7670?t=1730172622",
  "capsule_image": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/capsule_231x87.jpg?t=1730172622",
  "capsule_imagev5": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/capsule_184x69.jpg?t=1730172622",
  "categories": [
    {
      "description": "Um jogador",
      "id": 2,
    },
    {
      "description": "Compat. parcial com controle",
      "id": 18,
    },
    {
      "description": "Compartilhamento em família",
      "id": 62,
    },
  ],
  "detailed_description": "BioShock is a shooter unlike any you've ever played, loaded with weapons and tactics never seen. You'll have a complete arsenal at your disposal from simple revolvers to grenade launchers and chemical throwers, but you'll also be forced to genetically modify your DNA to create an even more deadly weapon: you. Injectable plasmids give you super human powers: blast electrical currents into water to electrocute multiple enemies, or freeze them solid and obliterate them with the swing of a wrench.  <br>No encounter ever plays out the same, and no two gamers will play the game the same way.<br><br><ul class="bb_ul"><li>Biologically modify your body: send fire storming from your fingertips and unleash a swarm of killer hornets hatched from the veins in your arms.<br></li><li>Hack devices and systems, upgrade your weapons and craft new ammo variants.<br></li><li>Turn everything into a weapon: the environment, your body, fire and water, and even your worst enemies.<br></li><li>Explore an incredible and unique art deco world hidden deep under the ocean.</li></ul><h2 class="bb_tag">BioShock: Breaking the Mold</h2>A free download that takes an inside look at the art of BioShock.",
  "developers": [
    "2K Boston",
    "2K Australia",
  ],
  "genres": [
    {
      "description": "Ação",
      "id": "1",
    },
    {
      "description": "RPG",
      "id": "3",
    },
  ],
  "header_image": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/header.jpg?t=1730172622",
  "is_free": false,
  "metacritic": {
    "score": 96,
    "url": "https://www.metacritic.com/game/pc/bioshock?ftag=MCD-06-10aaa1f",
  },
  "name": "BioShock™",
  "platforms": {
    "linux": false,
    "mac": false,
    "windows": true,
  },
  "publishers": [
    "2K",
  ],
  "release_date": {
    "coming_soon": false,
    "date": "21/ago./2007",
  },
  "required_age": 0,
  "screenshots": [
    {
      "id": 0,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002508.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002508.600x338.jpg?t=1730172622",
    },
    {
      "id": 1,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002509.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002509.600x338.jpg?t=1730172622",
    },
    {
      "id": 2,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002510.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002510.600x338.jpg?t=1730172622",
    },
    {
      "id": 3,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002511.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002511.600x338.jpg?t=1730172622",
    },
    {
      "id": 4,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002512.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002512.600x338.jpg?t=1730172622",
    },
    {
      "id": 5,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002513.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002513.600x338.jpg?t=1730172622",
    },
    {
      "id": 6,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002514.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002514.600x338.jpg?t=1730172622",
    },
    {
      "id": 7,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002515.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002515.600x338.jpg?t=1730172622",
    },
    {
      "id": 8,
      "path_full": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002517.1920x1080.jpg?t=1730172622",
      "path_thumbnail": "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/7670/0000002517.600x338.jpg?t=1730172622",
    },
  ],
  "short_description": "BioShock is a shooter unlike any you've ever played, loaded with weapons and tactics never seen. You'll have a complete arsenal at your disposal from simple revolvers to grenade launchers and chemical throwers, but you'll also be forced to genetically modify your DNA to create an even more deadly weapon: you.",
  "steam_appid": 7670,
  "type": "game",
  "website": "http://www.BioShockGame.com",
}
`;
